services/QuillLMS/spec/workers/copy_old_concept_results_to_concept_results_worker_spec.rb

```diff
@@ -5,14 +5,106 @@
+        expect(concept_result.concept_result_question_type.text).to eq(old_concept_result.question_type)
+      end
+
+      it 'should normalize empty string values as nil references to normalized tables' do
+        new_metadata = metadata.merge({'instructions': ''})
+        old_concept_result.update(metadata: new_metadata)
+
+        expect { subject.perform(old_concept_result.id, old_concept_result.id, true) }.to change(ConceptResult, :count).by(1)
+
+        concept_result = old_concept_result.concept_result
+
+        expect(concept_result.concept_result_instructions).to eq(nil)
+      end
+
+      it 'should be idempotent so that if the same ID is present twice, only one new record is created' do
+        expect do
+          subject.perform(old_concept_result.id, old_concept_result.id, true)
+          subject.perform(old_concept_result.id, old_concept_result.id, true)
+        end.to change(ConceptResult, :count).by(1)
       end
 
-      subject.perform(min_id, max_id)
+      it 'should skip items with IDs lower than start' do
+        old_concept_result
+        activity_session2 = create(:activity_session_without_concept_results, activity: activity)
+        used_old_concept_result = create(:sentence_combining, metadata: metadata, activity_session: activity_session2)
+
+        bulk_insert_worker_stub = double
+        expect(ConceptResult).to receive(:bulk_insert).and_yield(bulk_insert_worker_stub)
+
+        expect(bulk_insert_worker_stub).to receive(:add).with(hash_including(old_concept_result_id: used_old_concept_result.id))
+        expect(bulk_insert_worker_stub).not_to receive(:add).with(hash_including(old_concept_result_id: old_concept_result.id))
+
+        subject.perform(used_old_concept_result.id, used_old_concept_result.id, true)
+      end
+
+      it 'should not process items with IDs higher than finish' do
+        old_concept_result
+        activity_session2 = create(:activity_session_without_concept_results, activity: activity)
+        unused_old_concept_result = create(:sentence_combining, metadata: metadata, activity_session: activity_session2)
+
+        bulk_insert_worker_stub = double
+        expect(ConceptResult).to receive(:bulk_insert).and_yield(bulk_insert_worker_stub)
+
+        expect(bulk_insert_worker_stub).to receive(:add).with(hash_including(old_concept_result_id: old_concept_result.id))
+        expect(bulk_insert_worker_stub).not_to receive(:add).with(hash_including(old_concept_result_id: unused_old_concept_result.id))
+
+        subject.perform(old_concept_result.id, old_concept_result.id, true)
+      end
+
+      it 'should raise custom ConceptResultMigrationDeadlocked error when it triggers and ActiveRecord::Deadlocked error' do
+        expect(ConceptResult).to receive(:bulk_insert).and_raise(ActiveRecord::Deadlocked)
+
+        expect { subject.perform(1,1, true) }.to raise_error(CopyOldConceptResultsToConceptResultsWorker::ConceptResultMigrationDeadlocked)
+      end
+    end
```

Comment:
The contents of this `context` block is a very slightly modified (the third param that's required to pass to `perform`) version of all the tests before we modified it to do individual enqueuing.